src/Storages/ObjectStorageQueue/StorageObjectStorageQueue.cpp

```diff
@@ -578,69 +578,68 @@ bool StorageObjectStorageQueue::streamToViews()
 
     while (!shutdown_called && !file_iterator->isFinished())
     {
-        /// FIXME:
-        /// it is possible that MV is dropped just before we start the insert,
-        /// but in this case we would not throw any exception, so
-        /// data will not be inserted anywhere.
-        InterpreterInsertQuery interpreter(
-            insert,
-            queue_context,
-            /* allow_materialized */ false,
-            /* no_squash */ true,
-            /* no_destination */ true,
-            /* async_isnert */ false);
-        auto block_io = interpreter.execute();
-        auto read_from_format_info = prepareReadingFromFormat(
-            block_io.pipeline.getHeader().getNames(),
-            storage_snapshot,
-            queue_context,
-            supportsSubsetOfColumns(queue_context));
-
-        Pipes pipes;
+        QueryPipeline pipeline;
         std::vector<std::shared_ptr<ObjectStorageQueueSource>> sources;
-
-        pipes.reserve(processing_threads_num);
         sources.reserve(processing_threads_num);
 
         auto processing_progress = std::make_shared<ProcessingProgress>();
+
         for (size_t i = 0; i < processing_threads_num; ++i)
         {
+            /// FIXME:
+            /// it is possible that MV is dropped just before we start the insert,
+            /// but in this case we would not throw any exception, so
+            /// data will not be inserted anywhere.
+            InterpreterInsertQuery interpreter(
```

Comment:
This will not generate more data parts, but in case there are many small files in s3 - it would have been better to have a single insert to generate a single part, isn't it? Therefore, may be add a setting to control this?

Reply:
> I think it should be controlled with max_processing_* (i.e. max_processed_files_before_commit)

but this will anyway generate `min(processing_threads_num, max_processed_files_before_commit)` data parts while previously we could create 1 part containing all max_processed_files_before_commit files from s3. So it is not the same.

Reply:
You are right, but my point was that it does not make sense to use `processing_threads_num>1` if you don't need parallelism, and if you do, multiple INSERTs should be OK

But, if you still think that it make sense, then I'm fine with an extra setting. Any strong opinion on it's default then?

Reply:
>  if you don't need parallelism

this is a bit different, if you have 1000 files with 10 rows - it will be nice to read them in parallel with many threads but we still would want them to be written as 1 part if it is possible. So we do want parallelism but not for insert. Thus you get a very quick sync from s3, but without too many parts.

Reply:
Also a different thought - if we have several `InterpreterInsertQuery` objects, then it is possible that one/several of them could fail, while other do not. In this case - will it lead to rollback of all inserts or not? If not - then it will lead to incorrect commit of processed files, as currently code of `commit` method is not adjusted to this.

Reply:
Hm, good catch! And likely that was the reason that I have tons of duplicates in my tests

Reply:
Made it configurable with `parallel_inserts=true`.